feat: support `languageOptions.sourceType` (#1605)
web-infra-dev/rslint: architecture.md

accessors only when upstream behavior depends on provenance, instead of
rebuilding the merge. A rule whose upstream semantics add another source, such
as TypeScript library globals, applies this view last so `ecmaVersion` and
authored overrides remain authoritative. Non-global wrapper bindings remain a
`RefStore` initialization concern.

Before constructing rule contexts, the linter calls `ResolveLanguageDefaults`
once and passes its concrete `GlobalsInit`, `RefStoreInit`, and effective
`LanguageOptions` results to their respective consumers. An omitted source
type is filled from the filename for JavaScript files (`.cjs` → `commonjs`,

> `.js`/`.mjs` → `module`); `.ts`/`.tsx`/`.jsx` and other extensions keep the
> empty value. The resolver then selects inits from that effective source type:
> `commonjs` contributes writable `exports`, read-only `global`, `module`, and
> `require`, plus non-global wrapper scope and the wrapper-local `arguments`
> binding; `module` contributes a non-global top-level scope; `script` and the
> still-empty TypeScript/JSX value contribute no defaults. Authored `sourceType`
> therefore applies on every extension, including `.ts`/`.tsx`. The resolver
> does not inspect `package.json`. A rule reads
> `RuleContext.LanguageOptions` when its upstream behavior depends on them.

Reviewer: This paragraph predates the JS/TS split.

```suggestion
`commonjs` contributes writable `exports`, read-only `global`, `module`, and
`require` on every extension, plus — on espree-parsed extensions
(`.js`/`.jsx`/`.mjs`/`.cjs`) — non-global wrapper scope and the wrapper-local
`arguments` binding; `module` contributes a non-global top-level scope;
`script` and the still-empty TypeScript/JSX value contribute no defaults.
Authored `sourceType` therefore applies on every extension, including
`.ts`/`.tsx`. The resolver does not inspect `package.json`. A rule reads
`RuleContext.LanguageOptions` when its upstream behavior depends on them.
```

Author: Updated.
